### PR TITLE
Issue #1090 merge mf6hfb

### LINKED ISSUE #1090: Support combining HFB packages into a single HFB for MODFLOW 6
Related to #598 

Given that we require ``SingleLayerHorizontalFlowBarrierResistance`` only to have one layer and MODFLOW 6 requires a single HFB package, we need to be able to combine all HFB packages into a single one. 

Easiest I think is to combine the low-level ``Mf6HorizontalFlowBarrier`` packages into a single one.
Thing to investigate is if MODFLOW 6 supports values defined on a specific single cell edge multiple times, and if yes, what it does. Ideally MODFLOW6 would add resistances together.

This would need to occur in the ``Modflow6Model.write`` method.

### PR DESCRIPTION
**Pull request by JoerivanEngelen**
Fixes #1090

# Description
This changeset implements the following:
- Renames LayeredHorizontalFlowBarrier to SingleLayerHorizontalFlowBarrier 
- Moves duplictate logic into ``Package._from_dataset`` to construct package from a dataset
- Adds function to merge multiple HFBs into a low-level single hfb mf6 adapter
- Collect HFB packages into single list ``Model.write()`` and merge them with function
- ``_prepare_barrier_dataset_for_mf6_adapter`` function to prepare some required coordinate differences

# Checklist

- [x] Links to correct issue
- [x] Update changelog, if changes affect users
- [x] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [x] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

**Review by luitjansl — COMMENTED** [review]
looks good, nice tests.

File: imod/mf6/hfb.py [hunk 1]
```diff
@@ -273,6 +270,24 @@ def _vectorized_overlap(bounds_a: np.ndarray, bounds_b: np.ndarray):
     )
 
 
+def _prepare_barrier_dataset_for_mf6_adapter(dataset: xr.Dataset) -> xr.Dataset:
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/hfb.py:273]: Extended to:

>    Prepare barrier dataset for the initialization of Mf6HorizontalFlowBarrier.
>    The dataset is expected to have "edge_index" and "layer" coordinates and a
>   multi-index "cell_id" coordinate. The dataset contains as variables:
>    "cell_id1", "cell_id2", and "hydraulic_characteristic".

File: imod/mf6/hfb.py [hunk 2]
```diff
@@ -337,49 +352,34 @@ def _compute_barrier_values(
...
-            Grid with top of model layers.
-        bottom: GridDataArray
-            Grid with bottom of model layers.
-        k: GridDataArray
-            Grid with hydraulic conductivities.
-        validate: bool
-            Run validation before converting
+        Method does the following
+        - forces input grids to unstructured
+        - snaps lines to grid edges
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/hfb.py:365]: Done!

File: imod/mf6/hfb.py [hunk 3]
```diff
@@ -337,49 +352,34 @@ def _compute_barrier_values(
...
+        - removes invalid edges
+        - compute barrier values
+        - remove edge values to inactive cells
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/hfb.py:368]: Changed to:

>        Method does the following
>        - forces input grids to unstructured
>        - snaps lines to cell edges
>        - remove edge values connected to cell edges
>        - compute barrier values
>        - remove edge values to inactive cells
>        - finds connected cells in dataset

File: imod/mf6/hfb.py [hunk 4]
```diff
@@ -337,49 +352,34 @@ def _compute_barrier_values(
...
+        - finds connected cells in dataset
 
         Returns
         -------
-
+        dataset connected cells.
```
[the remark this answers is not in the document]
Comment (JoerivanEngelen) [reply imod/mf6/hfb.py:373]: Changed to

>        dataset with connected cells, containing:
>            - cell_id1
>            - cell_id2
>            - layer
>            - value name

**Pull request merged**